File: receipt-analyzer-service/src/main/java/pl/madzierski/daniel/app/receipt/revision/ReceiptRevisionEntity.java
package pl.madzierski.daniel.app.receipt.revision;

import jakarta.persistence.*;
import lombok.*;
import pl.madzierski.daniel.app.common.model.BaseEntity;
import pl.madzierski.daniel.app.receipt.ReceiptEntity;
import pl.madzierski.daniel.app.receipt.revision.item.ItemEntity;
import pl.madzierski.daniel.app.receipt.scan_resolver.ReceiptResolverStrategyType;

import java.util.Collection;
import java.util.HashSet;
import java.util.Objects;
import java.util.Set;
import java.util.stream.Collectors;

@Entity
@Table(name = "receipt_revision")
@Data
@NoArgsConstructor
@AllArgsConstructor
public class ReceiptRevisionEntity extends BaseEntity {

    @OneToMany(cascade = {CascadeType.ALL}, fetch = FetchType.LAZY, mappedBy = "receiptRevision")
    @Getter(AccessLevel.NONE)
    @Setter(AccessLevel.NONE)
    private final Set<ItemEntity> items = new HashSet<>();
    @OneToMany(mappedBy = "parentReceiptRevision")
    @Getter(AccessLevel.NONE)
    @Setter(AccessLevel.NONE)
    private final Set<ReceiptRevisionEntity> childReceiptRevisions = new HashSet<>();
    private String name;
    private String revision;
    private ReceiptResolverStrategyType resolver;
    private String brand;
    private Double totalPrice;
    private String payingDate;
    private String address;
    private Boolean isPreferredRevision;
    private Boolean isCorrect;
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "receipt_id")
    private ReceiptEntity receipt;
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "parent_receipt_revision_id")
    private ReceiptRevisionEntity parentReceiptRevision;

    public void addItem(ItemEntity item) {
        this.items.add(item);
    }

    public void addItems(Collection<ItemEntity> items) {
        this.items.addAll(items);
    }

    @Override
    public boolean equals(Object o) {
        if (o == null || getClass() != o.getClass()) return false;
        ReceiptRevisionEntity that = (ReceiptRevisionEntity) o;
        return Objects.equals(name, that.name) && Objects.equals(revision, that.revision) && resolver == that.resolver && Objects.equals(brand, that.brand) && Objects.equals(totalPrice, that.totalPrice) && Objects.equals(payingDate, that.payingDate) && Objects.equals(address, that.address) && Objects.equals(isPreferredRevision, that.isPreferredRevision) && Objects.equals(isCorrect, that.isCorrect);
    }

    @Override
    public int hashCode() {
        return Objects.hash(name, revision, resolver, brand, totalPrice, payingDate, address, isPreferredRevision, isCorrect);
    }

    public Set<ItemEntity> getItems() {
        return items.stream().collect(Collectors.toUnmodifiableSet());
    }

    public Boolean removeItem(String itemId) {
        return items.removeIf(itemEntity -> itemEntity.getId().equals(itemId));
    }

    public Set<ReceiptRevisionEntity> getChildReceiptRevisions() {
        return childReceiptRevisions.stream().collect(Collectors.toUnmodifiableSet());
    }
}

File: receipt-analyzer-service/src/main/kotlin/pl/madzierski/daniel/app/receipt/revision/ReceiptRevisionEntity.kt
package pl.madzierski.daniel.app.receipt.revision

import jakarta.persistence.*
import pl.madzierski.daniel.app.common.model.BaseEntity
import pl.madzierski.daniel.app.receipt.ReceiptEntity
import pl.madzierski.daniel.app.receipt.revision.item.ItemEntity
import pl.madzierski.daniel.app.receipt.scan_resolver.ReceiptResolverStrategyType

@Entity
@Table(name = "receipt_revision")
class ReceiptRevisionEntity(

    var name: String?,

    var revision: String?,

    var resolver: ReceiptResolverStrategyType?,

    var brand: String?,

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "receipt_id")
    var receipt: ReceiptEntity?,

    @OneToMany(mappedBy = "receiptRevision", fetch = FetchType.LAZY, cascade = [CascadeType.ALL])
    val items: MutableSet<ItemEntity> = mutableSetOf(),

    var totalPrice: Double?,

    var payingDate: String?,

    var address: String?,

    var isPreferredRevision: Boolean?,

    var isCorrect: Boolean?,

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "parent_receipt_revision_id")
    var parentReceiptRevision: ReceiptRevisionEntity? = null,

    @OneToMany(mappedBy = "parentReceiptRevision")
    val childReceiptRevisions: MutableSet<ReceiptRevisionEntity> = mutableSetOf()

) : BaseEntity() {

    fun addItem(item: ItemEntity) {
        item.receiptRevision = this
        items.add(item)
    }

    fun addItems(items: Collection<ItemEntity>) {
        items.forEach { it.receiptRevision = this }
        this.items.addAll(items)
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as ReceiptRevisionEntity

        if (totalPrice != other.totalPrice) return false
        if (isPreferredRevision != other.isPreferredRevision) return false
        if (isCorrect != other.isCorrect) return false
        if (revision != other.revision) return false
        if (resolver != other.resolver) return false
        if (brand != other.brand) return false
        if (items != other.items) return false
        if (payingDate != other.payingDate) return false
        if (address != other.address) return false

        return true
    }

    override fun hashCode(): Int {
        var result = totalPrice?.hashCode() ?: 0
        result = 31 * result + (isPreferredRevision?.hashCode() ?: 0)
        result = 31 * result + (isCorrect?.hashCode() ?: 0)
        result = 31 * result + (resolver?.hashCode() ?: 0)
        result = 31 * result + (brand?.hashCode() ?: 0)
        result = 31 * result + if (items != null) items.hashCode() else 0
        result = 31 * result + (payingDate?.hashCode() ?: 0)
        result = 31 * result + (address?.hashCode() ?: 0)
        return result
    }
}